File: am4/tests/game.rs
use am4::campaign::{Airline, Campaign, Eco};
// use am4::user::{GameMode, Role, User};

#[test]
fn test_campaign() {
    let c = Campaign::parse("c1, e");
    assert_eq!(c.pax_activated, Airline::C1_24HR);
    assert_eq!(c.cargo_activated, Airline::C1_24HR);
    assert_eq!(c.eco_activated, Eco::C24Hr);

    let c = Campaign::parse("e");
    assert_eq!(c.pax_activated, Airline::None);
    assert_eq!(c.cargo_activated, Airline::None);
    assert_eq!(c.eco_activated, Eco::C24Hr);
}

#[test]
fn test_campaign_reputation() {
    let c = Campaign::parse("c1, e");
    let rep = c.estimate_pax_reputation(45.0);
    assert_eq!(rep, 45.0 + 7.5 + 10.0);

    let c = Campaign::parse("c4, e");
    let rep = c.estimate_pax_reputation(45.0);
    assert_eq!(rep, 45.0 + 30.0 + 10.0);

    let c = Campaign::parse("e");
    let rep = c.estimate_pax_reputation(45.0);
    assert_eq!(rep, 45.0 + 10.0);
}

// #[test]
// fn test_default_user() {
//     let u = User::new(false);
//     let ur = User::new(true);

//     assert_eq!(u.game_mode, GameMode::Easy);
//     assert_eq!(ur.game_mode, GameMode::Realism);
//     assert_eq!(u.fuel_price, 700);
//     assert_eq!(u.co2_price, 120);
//     assert_eq!(u.load, 0.87);
//     assert_eq!(u.accumulated_count, 0);
//     assert_eq!(u.income_loss_tol, 0.02);
//     assert_eq!(u.fourx, false);
//     assert_eq!(u.role, Role::User);
// }

File: src/am4/db/models/game.py
from typing import Annotated, Literal

from pydantic import BaseModel, BeforeValidator, Field

from am4.utils.game import User

from .util import assert_equal_property_names

StrToUpper = BeforeValidator(func=lambda x: x.upper().strip() if isinstance(x, str) else x)
StrPctToFloat = BeforeValidator(
    func=lambda x: float(x.replace("%", "e-2")) if isinstance(x, str) and x.endswith("%") else x
)


class PyUser(BaseModel):
    id: str
    username: str
    game_id: Annotated[int, Field(gt=0)]
    game_name: str
    game_mode: Annotated[Literal["EASY", "REALISM"], StrToUpper]
    discord_id: Annotated[int, Field(ge=0)]
    wear_training: Annotated[int, Field(ge=0, le=5)]
    repair_training: Annotated[int, Field(ge=0, le=5)]
    l_training: Annotated[int, Field(ge=0, le=6)]
    h_training: Annotated[int, Field(ge=0, le=6)]
    fuel_training: Annotated[int, Field(ge=0, le=3)]
    co2_training: Annotated[int, Field(ge=0, le=5)]
    fuel_price: Annotated[int, Field(ge=0, le=3000)]
    co2_price: Annotated[int, Field(ge=0, le=200)]
    accumulated_count: Annotated[int, Field(ge=0)]
    load: Annotated[float, Field(ge=0, le=1), StrPctToFloat]
    income_loss_tol: Annotated[float, Field(ge=0, le=1), StrPctToFloat]
    fourx: bool
    role: Annotated[
        Literal[
            "BANNED",
            "USER",
            "TRUSTED_USER",
            "HIGHLY_TRUSTED_USER",
            "TOP_ALLIANCE_MEMBER",
            "TOP_ALLIANCE_ADMIN",
            "HELPER",
            "MODERATOR",
            "ADMIN",
            "SUPERUSER",
        ],
        StrToUpper,
    ]
    valid: bool


PyUserWhitelistedKeys = Literal[
    "game_id",
    "game_name",
    "game_mode",
    "wear_training",
    "repair_training",
    "l_training",
    "h_training",
    "fuel_training",
    "co2_training",
    "fuel_price",
    "co2_price",
    "accumulated_count",
    "load",
    "income_loss_tol",
    "fourx",
]

assert_equal_property_names(User, PyUser)
